packages/sdk/src/messageExtension/executeWithSSO.ts

```diff
@@ -0,0 +1,125 @@
+ * @param {TeamsFx} teamsfx - Used to provide configuration and auth.
+ * @param {string | string[]} scopes - The list of scopes for which the token will have access.
+ *
+ * @returns SignIn link CardAction with 200 status code.
+ */
+function getSignInResponseForMessageExtension(teamsfx: TeamsFx, scopes: string | string[]): any {
+  const scopesArray = getScopesArray(scopes);
+  const signInLink = `${teamsfx.getConfig("initiateLoginEndpoint")}?scope=${encodeURI(
+    scopesArray.join(" ")
+  )}&clientId=${teamsfx.getConfig("clientId")}&tenantId=${teamsfx.getConfig("tenantId")}`;
+  return {
+    composeExtension: {
+      type: "silentAuth",
+      suggestedActions: {
+        actions: [
+          {
+            type: "openUrl",
+            value: signInLink,
+            title: "Message Extension OAuth",
+          },
+        ],
+      },
+    },
+  };
+}
+
+/**
+ * execution in message extension with SSO token.
+ *
+ * @param {TurnContext} context - The context object for the current turn.
+ * @param {AuthenticationConfiguration} config - User custom the message extension authentication configuration.
+ * @param {string[]} scopes - The list of scopes for which the token will have access.
+ * @param {function} logic - The user execution code with SSO token.
+ *
+ * @throws {@link ErrorCode|InternalError} when failed to get access token with unknown error.
+ * @throws {@link ErrorCode|TokenExpiredError} when SSO token has already expired.
+ * @throws {@link ErrorCode|ServiceError} when failed to get access token from simple auth server.
+ * @throws {@link ErrorCode|InvalidParameter} when scopes is not a valid string or string array.
+ * @throws {@link ErrorCode|RuntimeNotSupported} when runtime is nodeJS.
+ *
+ * @returns A MessageExtension Response for the activity. If the logic not return any, return void instead.
+ */
+export async function executionWithToken(
+  context: TurnContext,
+  config: AuthenticationConfiguration,
+  scopes: string | string[],
+  logic?: (token: TeamsMsgExtTokenResponse) => Promise<any>
+): Promise<MessagingExtensionResponse | void> {
+  const valueObj = context.activity.value;
+  if (!valueObj.authentication || !valueObj.authentication.token) {
+    internalLogger.verbose("No AccessToken in request, return silentAuth for AccessToken");
+    return getSignInResponseForMessageExtension(new TeamsFx(undefined, config), scopes);
+  }
+  try {
+    const teamsfx = new TeamsFx(undefined, config).setSsoToken(valueObj.authentication.token);
+    const token: AccessToken | null = await teamsfx.getCredential().getToken(scopes);
+    const ssoTokenExpiration: number = parseJwt(valueObj.authentication.token).exp;
+    const tokenRes: TeamsMsgExtTokenResponse = {
+      ssoToken: valueObj.authentication.token,
+      ssoTokenExpiration: new Date(ssoTokenExpiration * 1000).toISOString(),
+      token: token!.token,
+      expiration: token!.expiresOnTimestamp.toString(),
+      connectionName: "",
+    };
+    if (logic) {
```

Comment:
Move this condition to the top of the function body?

Reply:
this logic parameter is optional, may be this API used in middleware which do not require any logic, but also need user doing the auth process. 
so put it here.

Reply:
Sure, I know